File: examples/sliders/sliders.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2024 Darby Johnston
// All rights reserved.

#include <tgUIApp/App.h>
#include <tgUIApp/Window.h>

#include <tgUI/GroupBox.h>
#include <tgUI/DoubleEditSlider.h>
#include <tgUI/FloatEditSlider.h>
#include <tgUI/IntEditSlider.h>
#include <tgUI/RowLayout.h>
#include <tgUI/ScrollWidget.h>

#include <tgCore/Format.h>

using namespace tg::core;
using namespace tg::ui;

TG_MAIN()
{
    try
    {
        auto context = Context::create();
        auto args = tg::app::convert(argc, argv);
        auto app = App::create(context, args, "sliders", "Sliders example");
        if (app->getExit() != 0)
            return app->getExit();

        // Create the window.
        auto window = Window::create(
            context,
            "sliders",
            Size2I(1280, 960));
        app->addWindow(window);

        // Create the layout.
        auto layout = VerticalLayout::create(context);
        layout->setMarginRole(SizeRole::Margin);
        auto scrollWidget = ScrollWidget::create(context, ScrollType::Both, window);
        scrollWidget->setWidget(layout);

        // Create integer sliders.
        auto groupBox = GroupBox::create(context, "Integer Sliders", layout);
        auto vLayout = VerticalLayout::create(context, groupBox);
        auto intEditSlider = IntEditSlider::create(context, nullptr, vLayout);
        intEditSlider->setCallback(
            [](int value)
            {
                std::cout << Format("Int: {0}").arg(value) << std::endl;
            });
        intEditSlider = IntEditSlider::create(context, nullptr, vLayout);
        intEditSlider->setRange(RangeI(100, 200));
        intEditSlider->setCallback(
            [](int value)
            {
                std::cout << Format("Int: {0}").arg(value) << std::endl;
            });
        intEditSlider = IntEditSlider::create(context, nullptr, vLayout);
        intEditSlider->setRange(RangeI(-100, 100));
        intEditSlider->setCallback(
            [](int value)
            {
                std::cout << Format("Int: {0}").arg(value) << std::endl;
            });

        // Create float sliders.
        groupBox = GroupBox::create(context, "Float Sliders", layout);
        vLayout = VerticalLayout::create(context, groupBox);
        auto floatEditSlider = FloatEditSlider::create(context, nullptr, vLayout);
        floatEditSlider->setCallback(
            [](float value)
            {
                std::cout << Format("Float: {0}").arg(value) << std::endl;
            });
        floatEditSlider = FloatEditSlider::create(context, nullptr, vLayout);
        floatEditSlider->setRange(RangeF(100.F, 200.F));
        floatEditSlider->setCallback(
            [](float value)
            {
                std::cout << Format("Float: {0}").arg(value) << std::endl;
            });
        floatEditSlider = FloatEditSlider::create(context, nullptr, vLayout);
        floatEditSlider->setRange(RangeF(-100.F, 100.F));
        floatEditSlider->setCallback(
            [](float value)
            {
                std::cout << Format("Float: {0}").arg(value) << std::endl;
            });

        // Create double sliders.
        groupBox = GroupBox::create(context, "Double Sliders", layout);
        vLayout = VerticalLayout::create(context, groupBox);
        auto doubleEditSlider = DoubleEditSlider::create(context, nullptr, vLayout);
        doubleEditSlider->setCallback(
            [](double value)
            {
                std::cout << Format("Double: {0}").arg(value) << std::endl;
            });
        doubleEditSlider = DoubleEditSlider::create(context, nullptr, vLayout);
        doubleEditSlider->setRange(RangeD(100.0, 200.0));
        doubleEditSlider->setCallback(
            [](double value)
            {
                std::cout << Format("Double: {0}").arg(value) << std::endl;
            });
        doubleEditSlider = DoubleEditSlider::create(context, nullptr, vLayout);
        doubleEditSlider->setRange(RangeD(-100.0, 100.0));
        doubleEditSlider->setCallback(
            [](double value)
            {
                std::cout << Format("Double: {0}").arg(value) << std::endl;
            });

        window->show();
        app->run();
    }
    catch (const std::exception& e)
    {
        std::cout << "ERROR: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}

File: python/examples/sliders/sliders.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2024 Darby Johnston
# All rights reserved.

from tinygfx import *
import sys

class Window(tgUIApp.Window):
    def __init__(self, context, name, size):
        tgUIApp.Window.__init__(self, context, name, size)
        
        # Create the layout.
        self.layout = tgUI.VerticalLayout(context)
        self.layout.marginRole = tgUI.SizeRole.Margin
        self.scrollWidget = tgUI.ScrollWidget(context, tgUI.ScrollType.Both, self)
        self.scrollWidget.setWidget(self.layout);
        
        # Create integer sliders.
        groupBox = tgUI.GroupBox(context, self.layout);
        groupBox.text = "Integer sliders"
        vLayout = tgUI.VerticalLayout(context, groupBox)
        intEditSlider = tgUI.IntEditSlider(context, None, vLayout)
        intEditSlider.setCallback(lambda value: print("Int:", value))
        intEditSlider = tgUI.IntEditSlider(context, None, vLayout)
        intEditSlider.range = tgCore.RangeI(100, 200)
        intEditSlider.setCallback(lambda value: print("Int:", value))
        intEditSlider = tgUI.IntEditSlider(context, None, vLayout)
        intEditSlider.range = tgCore.RangeI(-100, 100)
        intEditSlider.setCallback(lambda value: print("Int:", value))
        
        # Create float sliders.
        groupBox = tgUI.GroupBox(context, self.layout);
        groupBox.text = "Float sliders"
        vLayout = tgUI.VerticalLayout(context, groupBox)
        floatEditSlider = tgUI.FloatEditSlider(context, None, vLayout)
        floatEditSlider.setCallback(lambda value: print("Float:", value))
        floatEditSlider = tgUI.FloatEditSlider(context, None, vLayout)
        floatEditSlider.range = tgCore.RangeF(100, 200)
        floatEditSlider.setCallback(lambda value: print("Float:", value))
        floatEditSlider = tgUI.FloatEditSlider(context, None, vLayout)
        floatEditSlider.range = tgCore.RangeF(-100, 100)
        floatEditSlider.setCallback(lambda value: print("Float:", value))
        
        # Create double sliders.
        groupBox = tgUI.GroupBox(context, self.layout);
        groupBox.text = "Double sliders"
        vLayout = tgUI.VerticalLayout(context, groupBox)
        doubleEditSlider = tgUI.DoubleEditSlider(context, None, vLayout)
        doubleEditSlider.setCallback(lambda value: print("Double:", value))
        doubleEditSlider = tgUI.DoubleEditSlider(context, None, vLayout)
        doubleEditSlider.range = tgCore.RangeD(100, 200)
        doubleEditSlider.setCallback(lambda value: print("Double:", value))
        doubleEditSlider = tgUI.DoubleEditSlider(context, None, vLayout)
        doubleEditSlider.range = tgCore.RangeD(-100, 100)
        doubleEditSlider.setCallback(lambda value: print("Double:", value))
        
context = tgCore.Context()
app = tgUIApp.App(context, sys.argv, "buttons", "Buttons example")
if 0 == app.getExit():
    window = Window(context, "buttons", tgCore.Size2I(1280, 960))
    app.addWindow(window)
    window.show()
    app.run()
    window = None
